### ISSUE TITLE
Bad value assignment using list indices when type-promoting 

### ISSUE THREAD

**Issue by Balandat** [pytorch/pytorch#28010]
## 🐛 Bug

When type-promoting while assigning values using a list index, e.g. `A[:, [i]] = b`, where `A` and `b` are of different `dtype`, the assignment produces incorrect values.

## To Reproduce

```
A = torch.ones(5, 2, dtype=torch.double)
b = torch.rand(5) 
tensor([0.5433, 0.1091, 0.3717, 0.8436, 0.4544])


# Single-column assignment works fine
A[:, 1] = b
A
tensor([[1.0000, 0.5433],
        [1.0000, 0.1091],
        [1.0000, 0.3717],
        [1.0000, 0.8436],
        [1.0000, 0.4544]], dtype=torch.float64)


# Multi-column assignment breaks, presumably b/c of type promotion
A[:, [1]] = b.unsqueeze(-1)
A
tensor([[ 1.0000e+00,  1.1426e-10],
        [ 1.0000e+00,  1.8074e-06],
        [ 1.0000e+00,  1.4630e-03],
        [ 1.0000e+00,  1.1764e-05],
        [ 1.0000e+00, 5.2145e-315]], dtype=torch.float64)


# Multi-column assignment works if no type promotion is involved
A[:, [1]] = b.double().unsqueeze(-1)
A
tensor([[1.0000, 0.5433],
        [1.0000, 0.1091],
        [1.0000, 0.3717],
        [1.0000, 0.8436],
        [1.0000, 0.4544]], dtype=torch.float64)
```

## Expected behavior

Type promotion should work correctly and the second assignment above should assign correct values.

## Environment

pytorch master, linux, CPU (on fb infra)

cc @ezyang @gchanan @zou3519

**Comment by zou3519** [comment]
This is a correctness issue for a feature that is commonly used (indexing) so I am marking it as high priority.

**Comment by gchanan** [comment]
CC @nairbv.

**Comment by Balandat** [comment]
This is likely related to #26780

**Comment by nairbv** [comment]
also #27824

**Comment by ezyang** [comment]
Confirming that this repros still on master:

```
(/home/ezyang/Dev/pytorch-tmp-env) [ezyang@devgpu005.ash6 ~/Dev/pytorch-tmp] python test.py 
tensor([0.5321, 0.9535, 0.3994, 0.5588, 0.9358])
tensor([[1.0000, 0.5321],
        [1.0000, 0.9535],
        [1.0000, 0.3994],
        [1.0000, 0.5588],
        [1.0000, 0.9358]], dtype=torch.float64)
tensor([[ 1.0000e+00,  4.9087e-03],
        [ 1.0000e+00,  3.3975e-06],
        [ 1.0000e+00,  5.9220e-05],
        [ 1.0000e+00,  3.8529e-03],
        [ 1.0000e+00, 4.6585e-310]], dtype=torch.float64)
tensor([[1.0000, 0.5321],
        [1.0000, 0.9535],
        [1.0000, 0.3994],
        [1.0000, 0.5588],
        [1.0000, 0.9358]], dtype=torch.float64)
(/home/ezyang/Dev/pytorch-tmp-env) [ezyang@devgpu005.ash6 ~/Dev/pytorch-tmp] cat test.py
import torch

A = torch.ones(5, 2, dtype=torch.double)
b = torch.rand(5)
print(b)


# Single-column assignment works fine
A[:, 1] = b
print(A)


# Multi-column assignment breaks, presumably b/c of type promotion
A[:, [1]] = b.unsqueeze(-1)
print(A)


# Multi-column assignment works if no type promotion is involved
A[:, [1]] = b.double().unsqueeze(-1)
print(A)
```

**Comment by nairbv** [comment]
FYI this type of operation was an error in pytorch 1.2. The ability to perform this operation was not added intentionally.

e.g., Pytorch 1.2:
```
>>> a = torch.ones(5, 2, dtype=torch.double)
>>> b = torch.zeros(5, dtype=torch.float)
>>> a[:, [1]] = b.unsqueeze(-1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: expected device cpu and dtype Double but got device cpu and dtype Float
```

**Comment by gchanan** [comment]
It wasn't added intentionally because the intention was to only support cross-dtype operations on add, mul, sub, div?

**Comment by Balandat** [comment]
> FYI this type of operation was an error in pytorch 1.2. The ability to perform this operation was not added intentionally.

I guess the question is whether it makes sense for this operation to be supported going forward. Are there any downsides of supporting it?

**Comment by nairbv** [comment]
> It wasn't added intentionally because the intention was to only support cross-dtype operations on add, mul, sub, div?

That's right (plus the comparison ops)

>  guess the question is whether it makes sense for this operation to be supported going forward. Are there any downsides to supporting it?

Just the time it would take / prioritization afaik -- I'll probably re-add the error messages for this case in the meantime.

We'll have to fix #27824, and it seems that once that's fixed it may turn out to be trivial to add a correct version of this functionality.
